backend/uclapi/workspaces/occupeye.py

```diff
@@ -0,0 +1,741 @@
+                survey_id
+            )
+            survey_map_ids = self.r.lrange(
+                survey_map_ids_list,
+                0,
+                self.r.llen(survey_map_ids_list) - 1
+            )
+            survey_maps = []
+            for survey_map_id in survey_map_ids:
+                survey_map = self.r.hgetall(
+                    "occupeye:surveys:{}:maps:{}".format(
+                        survey_id,
+                        survey_map_id
+                    )
+                )
+                survey_maps.append(
+                    {
+                       "id": int(survey_map["id"]),
+                       "name": survey_map["name"],
+                       "image_id": int(survey_map["image_id"])
+                    }
+                )
+            survey["maps"] = survey_maps
+            surveys.append(survey)
+        return surveys
+
+    def _cache_image(self, image_id):
+        """
+        Downloads map images from the API and stores their
+        base64 representation and associated data type in Redis.
+        """
+        headers = {
+            "Authorization": self.get_bearer_token()
+        }
+        url = "{}/api/images/{}?deployment={}".format(
+            self.base_url,
+            image_id,
+            self.deployment_name
+        )
+        try:
+            request = requests.get(
+                url=url,
+                headers=headers,
+                stream=True
+            )
+            content_type = request.headers['Content-Type']
+        except:
+            raise BadOccupEyeRequest
+
+        raw_image = request.content
+        image_b64 = b64encode(raw_image)
+
+        pipeline = self.r.pipeline()
+        pipeline.set(
+            "occupeye:image:{}:base64".format(image_id),
+            image_b64
+        )
+        pipeline.expire(
+            "occupeye:image:{}:base64".format(image_id),
+            self.IMAGE_TTL
+        )
+        pipeline.set(
+            "occupeye:image:{}:content_type".format(image_id),
+            content_type
+        )
+        pipeline.expire(
+            "occupeye:image:{}:content_type".format(image_id),
+            self.IMAGE_TTL
+        )
+        pipeline.execute()
+
+    def get_image(self, image_id):
+        """
+        Pulls an Image ID requested by the user from Redis.
+        """
+        # We must ensure that only digits are passed into the code
+        # to prevent Redis injection attacks.
+        if not image_id.isdigit():
+            raise BadOccupEyeRequest
+
+        if not self.r.exists("occupeye:image:{}:base64".format(image_id)):
+            self._cache_image(image_id)
+
+        image_b64 = self.r.get("occupeye:image:{}:base64".format(image_id))
+        content_type = self.r.get(
+            "occupeye:image:{}:content_type".format(image_id)
+        )
+
+        return (image_b64, content_type)
+
+    def _get_survey_sensor_max_timestamp(self, survey_id):

```

Comment:
do we need to have leading underscores? It kinda makes the code a little harder to read imo

Reply:
Leading underscores denote internal functions that should not be called from outside of the class. It's the closest thing you can get to proper encapsulation, and it is [https://www.python.org/dev/peps/pep-0008/#method-names-and-instance-variables](respected by PEP8)